controlplane/kubeadm/internal/workload_cluster_coredns.go

```diff
@@ -195,6 +199,14 @@ func (w *Workload) getCoreDNSInfo(ctx context.Context, clusterConfig *bootstrapv
 	toImageRepository := parsedImage.Repository
 	if clusterConfig.ImageRepository != "" {
 		toImageRepository = strings.TrimSuffix(clusterConfig.ImageRepository, "/")
+	} else if kubernetesVersion.GTE(semver.Version{Major: 1, Minor: 25, Patch: 0, Pre: []semver.PRVersion{{VersionStr: "alpha"}}}) {
+		// Beginning with kubernetes v1.25, kubernetes images get published to registry.k8s.io instead of k8s.gcr.io.
+		// This ensures that the imageRepository gets patched to registry.k8s.io when upgrading from v1.24 or lower,
+		// but only if there was no imageRespository explicitly set at the KubeadmControlPlanes ClusterConfiguration.
+		// This follows the behavior of `kubeadm upgrade`.
+		if strings.HasPrefix(toImageRepository, oldKubernetesImageRepository) {
+			toImageRepository = KubernetesImageRepository + strings.TrimPrefix(toImageRepository, oldKubernetesImageRepository)
+		}
```

Comment:
I'm not sure this follow what is in the comment, we are not checking if imageRespository is explicitly set at the KubeadmControlPlanes ClusterConfiguration

Reply:
Makes now use of `ImageRepositoryFromClusterConfig` which has the wanted behaviour and is way more readable.